File: apps/browser-wallet/src/wallet/selection.ts
import type { SelectionResult, SpendCandidate } from "./models";

export function selectInputs(candidates: SpendCandidate[], targetValue: number): SelectionResult {
  if (targetValue <= 0) {
    throw new Error("Target value must be positive.");
  }
  const ordered = [...candidates].sort((left, right) => {
    if (left.amountChipbits !== right.amountChipbits) {
      return left.amountChipbits - right.amountChipbits;
    }
    if (left.txid !== right.txid) {
      return left.txid.localeCompare(right.txid);
    }
    return left.index - right.index;
  });

  const selected: SpendCandidate[] = [];
  let total = 0;
  for (const candidate of ordered) {
    selected.push(candidate);
    total += candidate.amountChipbits;
    if (total >= targetValue) {
      return {
        selected,
        totalInputChipbits: total,
        changeChipbits: total - targetValue,
      };
    }
  }
  throw new Error("Insufficient spendable balance for the requested amount and fee.");
}

File: src/chipcoin/wallet/selection.py
"""Wallet-side input selection strategies."""

from __future__ import annotations

from .models import SelectionResult, SpendCandidate


def select_inputs(candidates: list[SpendCandidate], target_value: int) -> SelectionResult:
    """Select spendable inputs for a target value."""

    if target_value <= 0:
        raise ValueError("Target value must be positive.")
    ordered = sorted(candidates, key=lambda candidate: (candidate.amount_chipbits, candidate.txid, candidate.index))
    selected: list[SpendCandidate] = []
    total = 0
    for candidate in ordered:
        selected.append(candidate)
        total += candidate.amount_chipbits
        if total >= target_value:
            return SelectionResult(
                selected=tuple(selected),
                total_input_chipbits=total,
                change_chipbits=total - target_value,
            )
    raise ValueError("Insufficient spendable balance for the requested amount and fee.")
